src/Stackage/Package/Git/Repository.hs

```diff
@@ -0,0 +1,352 @@
+-- created if `gitTagName` is `Nothing`, in which case `Nothing` is also
+-- returned.
+repoCreateTag :: GitRepository -- ^ Current repository
+              -> Ref -- ^ Commit reference.
+              -> ByteString -- ^ Tag message.
+              -> IO (Maybe Ref)
+repoCreateTag repo@GitRepository {repoInstance = GitInstance {..}
+                                 ,repoInfo = GitInfo {gitTagName = Just tagName
+                                                     ,..}} commitRef tagMessage = do
+  tag <- makeGitTag commitRef gitUser tagName tagMessage
+  newTag <-
+    case userGpgKey gitUser of
+      Just key -> signTag gitRepo key tag
+      Nothing -> return tag
+  ref <- repoWriteObject repo (Tag newTag)
+  writeFile (gitLocalPath </> "refs" </> "tags" </> S8.unpack (G.tagBlob newTag)) (show ref)
+  return $ Just ref
+repoCreateTag _ _ _ = return Nothing
+
+
+-- | Clones the repo if it doesn't exists locally yet, otherwise fetches and
+-- creates a local branch.
+ensureRepository
+  :: String -- ^ Git provider (e.g. "github.com")
+  -> String -- ^ Repository account (e.g. "commercialhaskell")
+  -> GitUser -- ^ User information to be used for commits and tags.
+  -> String -- ^ Repository name (e.g. "all-cabal-files")
+  -> String -- ^ Repository branch, make sure there is a remote branch with that
+            -- name. (e.g. "master")
+  -> FilePath -- ^ Location in the file system where repository should be cloned
+     -- to.
+  -> IO GitInfo
+ensureRepository repoHost repoAccount gitUser repoName repoBranchName repoBasePath = do
+  exists <- doesDirectoryExist repoLocalPath
+  if exists
+    then do
+      run repoLocalPath "git" ["remote", "set-url", "origin", repoAddress]
+      run
+        repoLocalPath
+        "git"
+        ["fetch", "origin", repoBranchName <> ":" <> repoBranchName]
+    else run
+           "."
+           "git"
+           [ "clone"
+           , "--bare"
+           , "--depth=1"
+           , repoAddress
+           , repoLocalPath
+           , "--branch"
+           , repoBranchName
+           ]
+  return
+    GitInfo
+    { gitAddress = repoAddress
+    , gitBranchName = repoBranchName
+    , gitUser = gitUser
+    , gitTagName = Nothing
+    , gitLocalPath = repoLocalPath
+    , gitRootTree = Nothing
+    }
+  where
+    repoLocalPath = repoBasePath </> repoName
+    repoAddress =
+      concat ["git@", repoHost, ":", repoAccount, "/", repoName, ".git"]
+
+
+-- | Signs a commit.
+signCommit :: Git -> String -> G.Commit -> IO G.Commit
+signCommit gitRepo key commit = do
+  signature <- signObject gitRepo key (G.ObjCommit commit)
+  -- Workaround for: https://github.com/vincenthz/hit/issues/35
+  -- Otherwise should be:
+  -- let signatureKey = "gpgsig"
+  --     signature = L.toStrict out
+  let signatureKey = L.toStrict $ L.append "gpgsig " $ L8.takeWhile (/='\n') signature
+      signature' = L.toStrict $ L.tail $ L8.dropWhile (/='\n') signature
```

Comment:
Yes, we know for sure that it will not be empty, if `signature` turns out to be empty or doesn't consist of many multiple lines it would be an error, different error, but still an error. So partiality here can be viewed as a good thing.